apps/labs/posts/libsf-error-state.md

```diff
@@ -0,0 +1,787 @@
+extension module, [`_ufuncs_cxx`](https://github.com/scipy/scipy/blob/main/scipy/special/meson.build#L187-L199) for
+scalar kernels written in C++. Through the byzantine machinations of SciPy special's build system, brought about by
+the need to handle scalar kernels written in 4 languages, the actual ufuncs from `_ufuncs_cxx` are defined in the
+`_ufuncs` extension module discussed earlier. Because `special.errstate` worked without issue for these ufuncs, and
+they appeared to be coming from the `_ufuncs` extension module, I never looked more deeply into `_ufuncs_cxx`. It
+turns out that `_ufuncs_cxx` has its own copy of `sf_error_actions` and it is updated in synchrony when setting the
+error state from Python.
+
+From [github.com/scipy/.../\_ufuncs_extra_code.pxi](https://github.com/scipy/scipy/blob/b98e79e4b326bf3beb4d861a744101ee7f438471/scipy/special/_ufuncs_extra_code.pxi#L158-L162):
+
+```python
+    for error, action in kwargs.items():
+        action = _sf_error_action_map[action]
+        code = _sf_error_code_map[error]
+        _set_action(code, action)
+        scipy.special._ufuncs_cxx._set_action(code, action)
+```
+
+I also found [an October 2016 comment on GitHub](https://github.com/scipy/scipy/pull/6722#issuecomment-256539142) from
+currently inactive SciPy maintainer Josh Wilson ([@person142](https://github.com/person142)), the original author of
+SciPy special's error handling features, which shows he was already aware of the problem we ended up facing. He was
+trying to make special function error handling work correctly in
+[`scipy.special.cython_special`](https://docs.scipy.org/doc/scipy/reference/special.cython_special.html), a module for
+calling scalar versions of special function from Cython. To quote:
+
+"To make the necessary modifications to cython_special error handling we need to instead share a single global variable
+between the two extension modules. On gcc we could do this with -rpath, but coming up with a portable solution seems
+messy. Is there a nice way out of this?"
+
+A _nice_ way? Perhaps not. He wisely chose instead to disable error handling in `cython_special` entirely. But maybe now
+that we at least have _a_ portable way to share global state between extension modules, this is worth revisiting.
+
+Before we finish, one more gem: in [another GitHub comment from
+2016](https://github.com/scipy/scipy/issues/6681#issuecomment-254192679), currently inactive SciPy BDFL Pauli Virtanen
+([@pv](https://github.com/pv)) suggested to "make `sf_error` error handling thread-local (which probably should be done
+in any case)". Over eight years later this is finally happening. I should probably go back and read old comment threads
+more often.
+
+### Funding acknowledgements
+
+This work was supported by the 2020 NASA ROSES grant, _Reinforcing the Foundations of Scientific Python_.
+
+---
+
+[^1]: The first regular shared library, not including Python extension modules.
```

Comment:
Also not including the vendored libraries? i.e. Everything in `scipy.libs`, `libscipy_openblas-*.so`, `libgfortran-*.so`.

Side note: I kinda expected that `libsf_error_state.so` would also be in `scipy.libs`.